python/data_server/dataset_cache.py

```diff
@@ -0,0 +1,42 @@
+import threading
+
+import cachetools
+
+from data_server import gcs_utils
+
+
+class DatasetCache():
+    def __init__(self, max_cache_size=8, cache_ttl=2 * 3600):
+        self.cache = cachetools.TTLCache(maxsize=max_cache_size, ttl=cache_ttl)
+        self.cache_lock = threading.Lock()
+
+    def clear(self):
+        """Clears entries from the cache. Mostly useful for tests."""
+        with self.cache_lock:
+            self.cache.clear()
+
+    def getDataset(self, gcs_bucket: str, table_id: str):
+        """Returns the given dataset identified by table_id as bytes.
+
+        getDataset will return the dataset from memory if it exists in the
+        cache. Otherwise, it will request the file from GCS and update the
+        cache on success.
+
+        gcs_bucket: Name of GCS bucket where the dataset is stored.
+        table_id: Name of the data set file to access.
+
+        Returns: Bytes object containing the dataset if successful. Throws
+        NotFoundError on failure."""
+        with self.cache_lock:
```

Comment:
It's been a while since I've written concurrent code so this may be a silly question, but do we need the lock? Setting a value in the cache seems like an atomic operation, and whether a reader gets the contents before or after it's updated doesn't really matter, right?

Reply:
The documentation for cachetools states that they are not thread-safe. Since we modify the cache we don't want to operate on the bare object. As an example, we could have a situation where there is 1 spot left in the cache and 2 threads try to write to it at the same time. Since they both see that there are 7 items in the cache, we end up leaking memory because it doesn't properly evict.

Reply:
Interesting, that makes sense. Does that apply for reading though? Could the lock just be around the writing?

I'm not sure that it will actually matter in practice because I'm guessing reading the cache will be fast, but my first thought when I saw this was I wonder if there would be performance issues of a bunch of connections trying to lock to read the cache at the same time.

Anyway, I leave this up to you

Reply:
Generally no, since it's still possible for a thread to try to read something that's being actively written. I'm not sure what the implementation of TTLCache is around this, but I don't think it's safe to assume that it is atomic.

I think that both reads and writes should be pretty fast, but this is something we will want to keep an eye on when we start receiving significant traffic.